### ISSUE TITLE
Atlas will occasionally fail to be generated

### ISSUE THREAD

**Issue by hanjoosten** [AmpersandTarski/RAP#102]
When the user loads a script and pushes the 'Atlas' button, an atlas is generated using the user's script. To do so, all elements in the script are taken from it, in terms of the formal ampersand model. I.e. all relations in a model are represented in the concept `relation`. In formal ampersand, there exist a relation `source[Relation * Concept]`. Every relation in the script of the user will make for a pair in that relation. There are a lot of relations in formal ampersand, and they are all populated based on the user's script. 

Formal Ampersand also has quite a few RULEs of it's own. The idea is, that whenever a valid script is taken to populate an atlas, that the population will not violate any invariants. This is pretty tricky however. We do not have a robust regression test in place. This makes that we currently cannot guarantee that for any script a valid population is being generated. If an invariant is violated, the violation will show as a red error. 

Any violations will be presented whenever an Atlas is being created:

![image](https://user-images.githubusercontent.com/1159672/101483427-718e7980-3958-11eb-9cc4-b3d85919ac00.png)

We need to get proper testing in place. 
Untill the time that we can quarentee that any valid script will lead to a working Atlas, the users of the Atlas button will be confronted with errors originating from formal ampersand.

**Comment by LloydRutledge** [comment]
Thanks for looking into this!

I get this problem running the Enrollment example at https://ampersandtarski.gitbook.io/documentation/tutorial-rap3/example-system-enrollment, even after fixing the isPartOf typing. The error is "Violation of rule InterfaceIdentity" with "Interface_1606926892_00553935[Interface],Interface_1607435420_00940582[Interface]
Interface_1607435420_00940582[Interface],Interface_1606926892_00553935[Interface]".

And when I click on the generated Atlas link I get "× Resource 'Context_1607435420_00936513' not found" and the rest of the display is empty. Is that the same problem?

I hold an open question/answer/discussion class with the 70 students this coming Thursday evening. Are there short term workarounds I can offer them about how to get atlas and/or prototype working for the tutorial example as well as their own projects? They are in week 4 of the 10 week cycle, with weekly readings in the online tutorial, and with the atlas and prototype being down the past several weeks, which they're eager to see. Any good news I can offer them Thursday about the server would be appreciated.

**Comment by hanjoosten** [comment]
This moring, Stef and I had a deep dive into one of the problems w.r.t. the atlas. As it turned out, there are several problems with the atlas. One of which is that multiple atlasses can be generated. The population of all those atlasses are likely to trigger rule violations. We found a way to clean the population before a new atlas is being generated. I hope @stefjoosten will be able to bring that fix into production soon (issue #96 ). That will solve a class of problems.

The prototype should work whenever a script can be succesfuly compiled. Obviously, as long as a script has errors, no prototype can be generated, which is fine. 

The atlas is currently more tricky. Whenever something is wrong with the extraction of population from the user's script into the formal ampersand model, one of the many rules could be violated. This wil prohibit the atlas to exist, because the transaction will be rolled back. As a result of that, the atom in the concept `Context` will not exist. If you click on the link that normally would lead you to the atlas, you indeed get a message  like `Resource 'Context_1607435420_00936513' not found` (the number might differ). It is unlikely that we will be able to solve this shortly, for we need a solid way to ensure that exactly the right population is extracted from any model. Only then we can be sure that that population never violates any invariant. 

So in short: The prototype should work. The atlas only works iff you are lucky 🍀 . But even if the atlas doesn't work, it could be educational to have a look at the error message. I.e. in your case, the violation is of rule InterfaceIdentity. Can you imagene what such a rule would do, if you think about this in the context of formal ampersand?

**Comment by LloydRutledge** [comment]
@hanjoosten your comment on making the prototype work is interesting. I'm setting a follow up question about that on issue #97 about the prototype.

**Comment by LloydRutledge** [comment]
I figured out how to stop those atlas red errors: you need to give a new unique name to INTERFACE - I'm guessing one no other saved script for any user has? That works for me with the Enrollment example from the tutorial. So I add gibberish after the I/F ID and now red atlas errors. The atlas generates fine. 

So I'll communicate this workaround to the students, that they need to add unique text after the INTERFACE name, until a fixed version of the server comes. Then they can run the atlas.

**Comment by stefjoosten** [comment]
# Diagnosis
The `IDENT` statement for interfaces was incorrect in `FormalAmpersand`. It did not take into account that an interface is uniquely determined by both its name and its context. This is fixed now.
